contentpool_client.module

```diff
@@ -45,6 +48,121 @@ function contentpool_client_form_remote_edit_form_alter(&$form, FormStateInterfa
+ *
+ * @param \Drupal\relaxed\Entity\RemoteInterface $remote
+ *   Remote interface.
+ *
+ * @return string[]
+ *   A list of render arrays which render treeselect filters.
+ */
+function contentpool_client_get_treeselect_filters(RemoteInterface $remote) {
+  if (!$remote->getThirdPartySetting('contentpool_client', 'is_contentpool', 0)) {
+    return [];
+  }
+
+  $termreference_fields = [];
+
+  $url = (string) $remote->uri();
+  // As the remote targets the relaxed endpoint we have to parse the url
+  // to get the base host.
+  $url_parts = parse_url($url);
+  $auth = [];
+  if (isset($url_parts['user']) && isset($url_parts['pass'])) {
+    $auth[] = $url_parts['user'];
+    $auth[] = $url_parts['pass'];
+  }
+  $base_url = $url_parts['scheme'] . '://' . $url_parts['host'];
+
+  try {
+    $response = (new Client())->get($base_url . '/api/contentpool-term-reference-fields?entity_type=node&bundle=article', [
+      'auth' => $auth,
+    ]);
+    if ($response->getStatusCode() == 200) {
+      $termreference_fields = json_decode($response->getBody()->getContents(), TRUE);
+    }
+  }
+  catch (\Exception $e) {
+    watchdog_exception('contentpool_client', $e);
+  }
+
+  if (empty($termreference_fields)) {
+    return [];
+  }
+
+  $treeselect_filters = [];
+  $current_filter = $remote->getThirdPartySetting('contentpool_client', 'filter', []);
+  foreach ($termreference_fields as $field => $field_data) {
+    $renderable = [
+      '#theme' => 'treeselect_filter',
+      '#field' => $field,
+      '#label' => $field_data['label'],
+      '#description' => t('Term reference filter: only sync content with selected %term.', ['%term' => $field_data['label']]),
+      '#attached' => [
+        'library' => [
+          'contentpool_client/treeselect_filter',
+        ],
+        'drupalSettings' => [
+          'contentpoolClient' => [
+            'filterData' => [
+              $field => [
+                'value' => isset($current_filter['node:article'][$field]) ? $current_filter['node:article'][$field] : NULL,
+                'options' => contentpool_client_map_treeselect_options($field_data['terms']),
+              ],
+            ],
+          ],
+        ],
+      ],
+    ];
+
+    $treeselect_filters[$field] = [
+      '#type' => 'markup',
+      '#markup' => \Drupal::service('renderer')->render($renderable),
+    ];
+  }
+
+  return $treeselect_filters;
+}
+
+/**
+ * Generate treeselect options array from the terms of the reference field.
+ *
+ * @param array $data
+ *   API result data of a reference field.
+ *
+ * @return array
+ *   Data array which is understood by the treeselect filter plugin.
+ */
+function contentpool_client_map_treeselect_options(array $data) {
+  $options = [];
+  foreach ($data as $item) {
+    $option = [
+      'id' => $item['id'],
+      'label' => $item['label'],
+    ];
+    if (!empty($item['children'])) {
+      $option['children'] = contentpool_client_map_treeselect_options($item['children']);
```

Comment:
does this actually do something. Seems unclear to me.

Reply:
Whatever the reason is, what be good to comment then.

Reply:
its a tree, so it should map the option recursively.